File: frontend/src/types/ticket.ts
import { UserDTO } from './user';

export interface TicketDTO {
    id: string;
    title: string;
    description: string;
    status: TicketStatus;
    priority: TicketPriority;
    reporter?: UserDTO;
    assignee?: UserDTO;
    createdAt: string;
    updatedAt: string;
}

export enum TicketStatus {
    OPEN = 'OPEN',
    IN_PROGRESS = 'IN_PROGRESS',
    CLOSED = 'CLOSED'
}

export enum TicketPriority {
    LOW = 'LOW',
    MEDIUM = 'MEDIUM',
    HIGH = 'HIGH',
    VERY_HIGH = 'VERY_HIGH'
}

export interface CreateTicketDTO {
    title: string;
    description: string;
    priority: TicketPriority;
    reporterId: string;
}

export interface UpdateTicketDTO {
    title: string;
    description: string;
    status: TicketStatus;
    priority: TicketPriority;
    reporterId?: string;
}

File: src/main/java/com/gnomeshift/tisk/ticket/Ticket.java
package com.gnomeshift.tisk.ticket;

import com.gnomeshift.tisk.user.User;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.Setter;
import lombok.ToString;
import lombok.NoArgsConstructor;
import lombok.Builder;
import org.hibernate.annotations.OnDelete;
import org.hibernate.annotations.OnDeleteAction;

import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "tickets")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@ToString(exclude = {"reporter", "assignee"})
public class Ticket {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @NotBlank
    @Size(min = 1, max = 255)
    private String title;

    @NotBlank
    @Size(min = 1, max = 10000)
    @Column(columnDefinition = "TEXT")
    private String description;

    @NotNull
    @Enumerated(EnumType.STRING)
    private TicketStatus status;

    @NotNull
    @Enumerated(EnumType.STRING)
    private TicketPriority priority;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "reporter_id")
    @OnDelete(action = OnDeleteAction.SET_NULL)
    private User reporter;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "assignee_id")
    @OnDelete(action = OnDeleteAction.SET_NULL)
    private User assignee;

    @NotNull
    private LocalDateTime createdAt;

    @NotNull
    private LocalDateTime updatedAt;

    private LocalDateTime closedAt;

    @PrePersist
    protected void onCreate() {
        this.createdAt = LocalDateTime.now();
        this.updatedAt = LocalDateTime.now();

        if (this.status == TicketStatus.CLOSED && this.closedAt == null) {
            this.closedAt = LocalDateTime.now();
        }
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = LocalDateTime.now();
    }

    public void updateStatusWithClosedAt(TicketStatus newStatus) {
        TicketStatus oldStatus = this.status;
        this.status = newStatus;

        if (newStatus == TicketStatus.CLOSED && oldStatus != TicketStatus.CLOSED) {
            this.closedAt = LocalDateTime.now();
        }
        else if (newStatus != TicketStatus.CLOSED && oldStatus == TicketStatus.CLOSED) {
            this.closedAt = null;
        }
    }
}
